### PR TITLE
feat: Add fallback positioning feature

### LINKED ISSUE #26874: Menu: Submenus should open above or below trigger if there is not enough horizontal space
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
any
```

### Are you reporting Accessibility issue?

None

### Reproduction

v9 storybook

## Actual Behavior
Submenu overflows available viewport

## Expected Behavior
Submenu should open above or below trigger if there is not enough horizontal space, similarly to v8 Menu:
image

### Logs

_No response_

### Requested priority

High

### Products/sites affected

Teams

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by ling1726**
Fixes #26874

Adds a `fallback` property to `usePositioning` hook that will be the position used if any overflow occurs after all positioning treatments.

Usage for the `Menu` component so that nested submenus will stack on top of each other when the viewport size is too small.

menu responsive

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/7bf995b3b64f3ffb07291fc6f7efedb00611f2de'>7bf995b3b64f3ffb07291fc6f7efedb00611f2de</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/Avatar-mount_base.html">1260</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/Avatar-mount.html">1292</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/Button-mount_base.html">919</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/Button-mount.html">918</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/Field-mount_base.html">1941</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/Field-mount.html">1951</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/FluentProvider-mount_base.html">1537</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/FluentProvider-mount.html">1552</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/FluentProviderWithTheme-mount_base.html">575</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/FluentProviderWithTheme-mount.html">575</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">540</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">539</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">572</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">574</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/InfoButton-mount_base.html">515</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/InfoButton-mount.html">520</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/MakeStyles-mount_base.html">1949</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/MakeStyles-mount.html">1981</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/Persona-mount_base.html">2868</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/Persona-mount.html">2809</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/SpinButton-mount_base.html">2277</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/26980/perf-test-react-components/SpinButton-mount.html">2306</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: packages/react-components/react-positioning/src/types.ts [hunk 1]
```diff
@@ -161,11 +161,27 @@ export interface PositioningOptions {
    */
   // eslint-disable-next-line @typescript-eslint/naming-convention
   unstable_disableTether?: boolean | 'all';
+
+  /**
+   * If flip fails to stop the positioned element from overflowing
+   * its boundaries, use a specified fallback positions.
+   */
+  fallback?: PositioningShorthandValue[];
```
Comment (bsunderhus) [inline packages/react-components/react-positioning/src/types.ts:169]:
```suggestion
  fallbacks?: PositioningShorthandValue[];
```

File: packages/react-components/react-positioning/src/usePositioning.ts [hunk 2]
```diff
@@ -150,7 +150,7 @@ export function usePositioning(options: UsePositioningOptions): UsePositioningRe
   return { targetRef: setTarget, containerRef: setContainer, arrowRef: setArrow };
 }
 
-interface UsePositioningOptions extends PositioningProps {
+interface UsePositioningOptions extends PositioningProps, Pick<PositioningOptions, 'fallback' | 'pinned'> {
```
Comment (bsunderhus) [inline packages/react-components/react-positioning/src/usePositioning.ts:153]:
```suggestion
interface UsePositioningOptions extends PositioningProps, Pick<PositioningOptions, 'fallbacks' | 'pinned'> {
```

File: packages/react-components/react-positioning/src/usePositioning.ts [hunk 3]
```diff
@@ -171,6 +171,7 @@ function usePositioningOptions(options: PositioningOptions) {
     unstable_disableTether: disableTether,
     positionFixed,
     overflowBoundaryPadding,
+    fallback,
```
Comment (bsunderhus) [inline packages/react-components/react-positioning/src/usePositioning.ts:174]:
```suggestion
    fallbacks,
```

File: packages/react-components/react-positioning/src/usePositioning.ts [hunk 4]
```diff
@@ -184,7 +185,7 @@ function usePositioningOptions(options: PositioningOptions) {
       const middleware = [
         offset && offsetMiddleware(offset),
         coverTarget && coverTargetMiddleware(),
-        !pinned && flipMiddleware({ container, flipBoundary, hasScrollableElement }),
+        !pinned && flipMiddleware({ container, flipBoundary, hasScrollableElement, isRtl, fallback }),
```
Comment (bsunderhus) [inline packages/react-components/react-positioning/src/usePositioning.ts:188]:
```suggestion
        !pinned && flipMiddleware({ container, flipBoundary, hasScrollableElement, isRtl, fallbacks }),
```

File: packages/react-components/react-positioning/src/usePositioning.ts [hunk 5]
```diff
@@ -222,6 +223,7 @@ function usePositioningOptions(options: PositioningOptions) {
       position,
       strategy,
       overflowBoundaryPadding,
+      fallback,
```
Comment (bsunderhus) [inline packages/react-components/react-positioning/src/usePositioning.ts:226]:
```suggestion
      fallbacks,
```

**Review by bsunderhus — APPROVED** [review]
nit: `fallbacks` >>>> `fallback`

LGTM

File: packages/react-components/react-positioning/src/middleware/flip.ts [hunk 6]
```diff
@@ -1,18 +1,27 @@
...
 
 export function flip(options: FlipMiddlewareOptions) {
-  const { hasScrollableElement, flipBoundary, container } = options;
+  const { hasScrollableElement, flipBoundary, container, fallback = [], isRtl } = options;
+  const fallbackPlacements = fallback
+    .map(shorthand => {
+      const { position, align } = resolvePositioningShorthand(shorthand);
+      return toFloatingUIPlacement(align, position, isRtl);
+    })
+    .filter(Boolean) as Placement[];
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/react-components/react-positioning/src/middleware/flip.ts]: renamed to `fallbackPositions`

File: packages/react-components/react-menu/src/components/Menu/useMenu.tsx [hunk 7]
```diff
@@ -43,6 +59,7 @@ export const useMenu_unstable = (props: MenuProps): MenuState => {
     position: isSubmenu ? 'after' : 'below',
     align: isSubmenu ? 'top' : 'start',
     target: props.openOnContext ? contextTarget : undefined,
+    fallback: isSubmenu ? submenuFallbackPositions : undefined,
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/react-components/react-menu/src/components/Menu/useMenu.tsx]: renamed to `fallbackPositions`

File: packages/react-components/react-positioning/src/types.ts (hunk 1 above)
Comment (ling1726) [reply packages/react-components/react-positioning/src/types.ts]: renamed to `fallbackPositions`

File: packages/react-components/react-positioning/src/usePositioning.ts (hunk 3 above)
Comment (ling1726) [reply packages/react-components/react-positioning/src/usePositioning.ts]: renamed to `fallbackPositions`

File: packages/react-components/react-positioning/src/usePositioning.ts (hunk 4 above)
Comment (ling1726) [reply packages/react-components/react-positioning/src/usePositioning.ts]: renamed to `fallbackPositions`

File: packages/react-components/react-positioning/src/usePositioning.ts (hunk 2 above)
Comment (ling1726) [reply packages/react-components/react-positioning/src/usePositioning.ts]: renamed to `fallbackPositions`

File: apps/vr-tests-react-components/src/stories/Menu/NestedMenuSmallViewportFlipped.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
+import * as React from 'react';
+
+import { Menu, MenuTrigger, MenuList, MenuItem, MenuPopover } from '@fluentui/react-menu';
+import { makeStyles, shorthands } from '@griffel/react';
+import { PositioningImperativeRef } from '@fluentui/react-positioning';
+import { withStoryWrightSteps } from '../../utilities/withStoryWrightSteps';
```
Comment (ling1726) [inline apps/vr-tests-react-components/src/stories/Menu/NestedMenuSmallViewportFlipped.stories.tsx:6]:
```suggestion
```

File: apps/vr-tests-react-components/src/stories/Menu/NestedMenuSmallViewportStacked.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,78 @@
+import * as React from 'react';
+
+import { Menu, MenuTrigger, MenuList, MenuItem, MenuPopover } from '@fluentui/react-menu';
+import { makeStyles, shorthands } from '@griffel/react';
+import { PositioningImperativeRef } from '@fluentui/react-positioning';
+import { withStoryWrightSteps } from '../../utilities/withStoryWrightSteps';
```
Comment (ling1726) [inline apps/vr-tests-react-components/src/stories/Menu/NestedMenuSmallViewportStacked.stories.tsx:6]:
```suggestion
```

**Pull request merged**